spec/02-identifiers-names-and-scopes.md

```diff
@@ -17,14 +17,16 @@ which are collectively called _bindings_.
 Bindings of each kind are assigned a precedence which determines
 whether one binding can shadow another:
 
-1. Definitions and declarations that are local, inherited, or made
-   available by a package clause and also defined in the same compilation unit
-   as the reference to them, have the highest precedence.
+1. Definitions and declarations in lexical scope that are not [top-level](09-top-level-definitions.html)
```

Comment:
@som-snytt I assume `top-level` is excluded here because that's what's implemented in Scala 3 (https://github.com/lampepfl/dotty/pull/8622). The following compiles and `new X` resolves to `C.this.X`

```
package p {
  trait T { class X }
  class C extends T {
    def test = new X
  }
  class X
}
```

Changing `package p` to `object p` makes `new X` ambiguous. Do you see a good reason for that? cc @odersky

Reply:
If `T` has a member `def name: String = "John Doe"`, then just where it can be inherited, introducing a different definition of `name` might be in conflict.

Packagings are not classes and don't suffer from extension, so I should be able to write top-level `def name = "Gloria Mundi"` without concern about confusing anyone.

But this is ambiguous:
```
    trait T {
      class X
      def name: String = "John Doe"
    }
    class C extends T {
      def test = new X
      def result = name
    }
    class X
    //def name = "Gloria Mundi"
    val name: T = new T {
      def result = name
    }
```
I expected it to prefer the inherited `name` and not the enclosing definition which is a top-level def.
```
15 |      def result = name
   |                   ^^^^
   |                   Reference to name is ambiguous,
   |                   it is both defined in package object t10220$package
   |                   and inherited subsequently in anonymous class Object with p.T {...}
```
I don't think they actually got rid of package objects...

Reply:
So you expect an inherited symbol takes precedence over an outer definition in the same source file, if that outer definition is package-owned? Can you say why? I don't yet see what difference it makes if the outer symbol is top-level or not.

> I should be able to write top-level `def name = "Gloria Mundi"` without concern about confusing anyone

What is the difference in that regard between `package p { def name = "..." }` and `object p { def name = "..."}`?

Reply:
For the record, I'm on this topic because I'm testing https://github.com/scala/scala/pull/10297, and I see warnings that didn't show up in 2.13. These new warnings seem helpful to me.

Reply:
I guess the proof is in the pudding, or in golf it's the putting.

I have only a rationalization: in `package p`, I know nothing is inherited in this scope (because we do not `package object p extends T`). Once someone defines a class, the folks working in that mine must manage that namespace, where stuff is inherited.

I agree there is regularity in seeing `object p` as just a way to "associate" some definitions, and I think they speak of "package membership", but if the regularity were borne out, then we would love `package object p extends T`. You would be asking, Why is that different from `object p extends T`?

But all that is just a rationalization.

What are some examples of pernicious shadowing of top-level defs? I can't wait to see.

Reply:
An example was:

```scala
package p

class Code extends Parent {
  def impl = s"The $CommonName"
}

case class CommonName(some: Thing)
```

A second file:

```scala
package p
trait Parent {
  val CommonName = "something"
}
```

It's maybe clear here that we're not intending to refer to the synthetic case class companion, but this could still be a head scratcher for someone looking at the first file. One could for example do a search in the editor for `CommonName` and land on the case class.

In this case I moved `CommonName` to a separate file instead of changing the reference.

Reply:
Aha, that does seem lintable, at least with `CommonName` in the same file. Conversely, it loses precedence when moved to a different file, but the reference in `Parent` does not.

I just realized this feature is not released until 2.13.11. Let me take a moment to clear my fever-addled brain. 

I notice `toString` is not exempt.

Also I was about to wonder if the use case is really anon classes, esp SAM types. The user expectation is more obviously, "I am focused on implementing that one method, and probably unaware of inheriting anything."

Reply:
> `CommonName` in the same file

Yes, I mean to warn only if the outer definition is in the same compilation unit.

> I notice `toString` is not exempt.
> if the use case is really anon classes

Those are very good points. I certainly saw it pop up a lot with anonymous classes, and one would certainly expect `toString` to resolve in the current object. Hmm. Since this is a spec change and not an impl detail (a lint rule), we need to have general and hopefully simple rules. An exception for `toString` wouldn't work...

Reply:
I think we can wait to see if `toString` really is problematic. So far (Scala 3, my tests of the new Scala 2 warning) it didn't seem so.